rootfs/scheduler/__init__.py

```diff
@@ -1,34 +1,784 @@
+            self._create_namespace(app_name)
+            self._create_secret(app_name)
+            self._create_rc(name, image, command, **kwargs)
+            self._create_service(name, app_name, app_type, data, image=image)
+        except Exception as e:
+            logger.debug(e)
+            self._scale_rc(name, 0, app_name)
+            self._delete_rc(name, app_name)
+            raise
+
+    def start(self, name):
+        """Start a container."""
+        pass
+
+    def stop(self, name):
+        """Stop a container."""
+        pass
 
     def destroy(self, name):
-        """Destroy a container."""
-        raise NotImplementedError
+        """Destroy a application by deleting its namespace."""
+        namespace = name.split("_")[0]
+        logger.debug("destroy {}".format(name))
+        url = self._api("/namespaces/{}", namespace)
+        resp = self.session.delete(url)
+        if resp.status_code == 404:
+            logger.warn('delete Namespace "{}": not found'.format(namespace))
+        elif resp.status_code != 200:
+            error(resp, 'delete Namespace "{}"', namespace)
+
+    def logs(self, name):
+        logger.debug("logs {}".format(name))
+        app_name = name.split('_')[0]
+        name = name.replace('.', '-').replace('_', '-')
+        status, data, reason = self._get_pods(app_name)
+        parsed_json = json.loads(data)
+        log_data = ''
+        for pod in parsed_json['items']:
+            if name in pod['metadata']['generateName'] and pod['status']['phase'] == 'Running':
+                status, data, reason = self._pod_log(pod['metadata']['name'], app_name)
+                log_data += data
+
+        return log_data
 
     def run(self, name, image, entrypoint, command):
         """Run a one-off command."""
-        raise NotImplementedError
+        logger.debug('run {}, img {}, entypoint {}, cmd "{}"'.format(
+            name, image, entrypoint, command))
+        appname = name.split('_')[0]
+        name = name.replace('.', '-').replace('_', '-')
+        l = {
+            'id': name,
+            'version': self.apiversion,
+            'image': self.registry + '/' + image,
+        }
 
-    def start(self, name):
-        """Start a container."""
-        raise NotImplementedError
+        template = string.Template(POD_TEMPLATE).substitute(l)
+        if command.startswith('-c '):
+            args = command.split(' ', 1)
+            args[1] = args[1][1:-1]
+        else:
+            args = [command[1:-1]]
+
+        js_template = json.loads(template)
+        js_template['spec']['containers'][0]['command'] = [entrypoint]
+        js_template['spec']['containers'][0]['args'] = args
+        url = self._api("/namespaces/{}/pods", appname)
+        resp = self.session.post(url, json=js_template)
+        if unhealthy(resp.status_code):
+            error(resp, 'create Pod in Namespace "{}"', appname)
+
+        while(1):
```

Comment:
The code is moved from elsewhere but you do make a good point:

It would continue on based on what I am reading, https://github.com/kubernetes/kubernetes/blob/master/docs/user-guide/pod-states.md#pod-phase lists out the phases

If you want to create a ticket for this then that'd be rad - we could def write this to be more resilient